Make the histogram analyzer work on replayed cw2dmk logs too

TimothyPMann opened the pull request:
Also rename the binary from cwtsthst to cwhisto.  The "test" part of the original name (testhist) doesn't have any meaning, so let's drop it from the name.

TimothyPMann committed:
Make the histogram analyzer work on replayed cw2dmk logs too

Also rename the binary from cwtsthst to cwhisto.  The "test" part of
the original name (testhist) doesn't have any meaning, so let's drop
it from the name.

TimothyPMann commented:
I haven't yet regression tested to make sure this still works for direct reading from the catweasel. So not quite ready to merge.

TimothyPMann committed:
Rename testhist.c to cwhisto.c

TimothyPMann committed:
Rework cwhisto command line parsing

Use option letters (instead of having all parameters be positional and
most mandatory).  This prepares for adding more options.

To do: More thorough checking for invalid option settings.

TimothyPMann commented:
Where I'm going with this branch: I would like to add some code to optionally show the histograms with gnuplot instead of as just a retangular blob of 128 numbers. This is pretty easy except for some details, like deciding what to do in the replay case, where you might have as many as 160 histograms  (80 tracks * 2 sides) even with no retries or with split = 0.

qbarnes commented:
You might want to take a look at the GW rewrite, `gwhist`, its command line interface (https://github.com/qbarnes/gw2dmk/blob/master/gwhist.c#L103-L107), and its man page (https://github.com/qbarnes/gw2dmk/blob/master/gwhist.1) to see what ideas you might want to copy.  (In looking at it myself just now, I found already a few typos and corrections I need to make.  Ugh!  :) )

In the rewrite, it does have the `-r` option to tell how many revolutions to sample a g histogram and merge/average the data of a given track and side.

I never included multiple track/side reads because it's easy enough to just put a shell for loop around it and call it as many times as needed with the changing track and sides specified on the command line and collect the data into different files or one file.

The gnuplot option sounds cool, but does it need to be built in?  Can't it just be done in a post-processing step?

One thing I'd like is an command line option that would spit out the raw data in JSON format so the data can be read and post-processed by any tool.

TimothyPMann commented:
Nice, I'll take a look. I think you're right that just a way to get the raw
data out in an easily parsable form is what's most needed. Maybe also a
sample script for gnuplot.

On Mon, Jun 2, 2025, 10:46 AM Quentin Barnes ***@***.***>
wrote:

> *qbarnes* left a comment (qbarnes/cw2dmk#30)
> <https://github.com/qbarnes/cw2dmk/pull/30#issuecomment-2931768282>
>
> You might want to take a look at the GW rewrite, gwhist, its command line
> interface (
> https://github.com/qbarnes/gw2dmk/blob/master/gwhist.c#L103-L107), and
> its man page (https://github.com/qbarnes/gw2dmk/blob/master/gwhist.1) to
> see what ideas you might want to copy. (In looking at it myself just now, I
> found already a few typos and corrections I need to make. Ugh! :) )
>
> In the rewrite, it does have the -r option to tell how many revolutions
> to sample a g histogram and merge/average the data of a given track and
> side.
>
> I never included multiple track/side reads because it's easy enough to
> just put a shell for loop around it and call it as many times as needed
> with the changing track and sides specified on the command line and collect
> the data into different files or one file.
>
> The gnuplot option sounds cool, but does it need to be built in? Can't it
> just be done in a post-processing step?
>
> One thing I'd like is an command line option that would spit out the raw
> data in JSON format so the data can be read and post-processed by any tool.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qbarnes/cw2dmk/pull/30#issuecomment-2931768282>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AILDEHIEDIWGCJ6RCL7WJ5D3BSEV5AVCNFSM6AAAAAB56WN742VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMZRG43DQMRYGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TimothyPMann committed:
Let's call it cwhist instead of cwhisto

TimothyPMann committed:
Add an option to specify the number of passes

TimothyPMann committed:
Add -H feature to support plotting the histograms with gnuplot

Also some cleanup and minor fixes.  In particular, make the minimum
count in a bucket for peak detection depend on the number of passes,
so that using a large number of passes doesn't lead to noise causing
false peaks to be detected.

TimothyPMann commented:
> I haven't yet regression tested to make sure this still works for direct reading from the catweasel. So not quite ready to merge.

Direct reading from catweasel is tested now too, as well as the other new features.

TimothyPMann committed:
Mild cleanup to split computation from printing
